include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -3,428 +3,573 @@
-template<>\
-struct find_lsb_helper<INPUT_INTEGER_TYPE>\
-{\
-	static int32_t findLSB(NBL_CONST_REF_ARG(INPUT_INTEGER_TYPE) val)\
-	{\
-		return find_lsb_helper<INTEGER_TYPE>::findLSB(val);\
-	}\
-};\
+template<typename Vectorial>
+NBL_PARTIAL_REQ_TOP(concepts::Vectorial<Vectorial>)
+struct dot_helper<Vectorial NBL_PARTIAL_REQ_BOT(concepts::Vectorial<Vectorial>) >
+{
+	using scalar_type = typename vector_traits<Vectorial>::scalar_type;
 
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(int16_t, int32_t)
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(uint16_t, uint32_t)
-#ifndef __HLSL_VERSION
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(int8_t, int32_t)
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(uint8_t, uint32_t)
-#endif
+	static inline scalar_type __call(NBL_CONST_REF_ARG(Vectorial) lhs, NBL_CONST_REF_ARG(Vectorial) rhs)
+	{
+		static const uint32_t ArrayDim = vector_traits<Vectorial>::Dimension;
+		static array_get<Vectorial, scalar_type> getter;
 
-template<>
-struct find_lsb_helper<uint64_t>
+		scalar_type retval = getter(lhs, 0) * getter(rhs, 0);
+		for (uint32_t i = 1; i < ArrayDim; ++i)
+			retval = retval + getter(lhs, i) * getter(rhs, i);
+
+		return retval;
+	}
+};
+template<typename FloatingPointLikeVectorial>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointLikeVectorial<FloatingPointLikeVectorial> && (vector_traits<FloatingPointLikeVectorial>::Dimension == 3))
+struct cross_helper<FloatingPointLikeVectorial NBL_PARTIAL_REQ_BOT(concepts::FloatingPointLikeVectorial<FloatingPointLikeVectorial> && (vector_traits<FloatingPointLikeVectorial>::Dimension == 3)) >
 {
-	static int32_t findLSB(NBL_CONST_REF_ARG(uint64_t) val)
+	static FloatingPointLikeVectorial __call(NBL_CONST_REF_ARG(FloatingPointLikeVectorial) lhs, NBL_CONST_REF_ARG(FloatingPointLikeVectorial) rhs)
 	{
 #ifdef __HLSL_VERSION
-		const uint32_t lowBits = uint32_t(val);
-		const int32_t lowLsb = find_lsb_helper<uint32_t>::findLSB(lowBits);
+		if (hlsl::is_vector_v<FloatingPointLikeVectorial>)
+			return spirv::cross(lhs, rhs);
+#else
+		using traits = hlsl::vector_traits<FloatingPointLikeVectorial>;
+		array_get<FloatingPointLikeVectorial, typename traits::scalar_type> getter;
+		array_set<FloatingPointLikeVectorial, typename traits::scalar_type> setter;
 
-		if (lowLsb == -1)
-		{
-			const uint32_t highBits = uint32_t(val >> 32);
-			const int32_t highLsb = find_lsb_helper<uint32_t>::findLSB(highBits);
-			if (highLsb == -1)
-				return -1;
-			else
-				return 32 + highLsb;
-		}
+		FloatingPointLikeVectorial output;
+		setter(output, 0, getter(lhs, 1) * getter(rhs, 2) - getter(rhs, 1) * getter(lhs, 2));
+		setter(output, 1, getter(lhs, 2) * getter(rhs, 0) - getter(rhs, 2) * getter(lhs, 0));
+		setter(output, 2, getter(lhs, 0) * getter(rhs, 1) - getter(rhs, 0) * getter(lhs, 1));
 
-		return lowLsb;
-#else
-		return glm::findLSB(val);
+		return output;
 #endif
 	}
 };
 
-template<int N>
-struct find_lsb_helper<vector<int32_t, N> >
+template<typename Vector>
+NBL_PARTIAL_REQ_TOP(is_vector_v<Vector>)
```

Comment:
why not any vectorial?

Reply:
I guess same question can be asked for every helper that follows:
- bitReverse
- min
- max

Reply:
outdated i guess